### PR TITLE
Thread Pool Job Executor

### PR DESCRIPTION
**Pull request by dleehr**
Implements a JobExecutor using [concurrent.futures.ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor)

Does not implement any kubernetes-specific behavior, but provides a base class to simplify the parallel executor and be easily extensible.

### CODE REVIEW

**Comment by dleehr** [comment]
Not ready to merge, just wanted to get some code review :)

**Comment by dleehr** [comment]
Now connected to main.py and cleaned up. ready for re-review

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.6
+FROM python:3.7
```
Comment (johnbradley) [inline Dockerfile]: Do we need to upgrade the conformance dockerfile to use 3.7 as well?
https://github.com/Duke-GCB/calrissian/blob/2ed785497852f144da8236caccfedad556802d98/conformance/Dockerfile.conformance#L1

File: calrissian/executor.py [hunk 2]
```diff
@@ -1,40 +1,416 @@
...
+            # Note that cancel will only matter if there aren't enough available threads to start processing the job in
+            # the first place. Once the function starts running it cannot be cancelled.
+            logger.error('Found a queued exception, canceling outstanding futures')
+            for f in futures:
+                f.cancel()
+            # Wait for outstanding futures to finish up so that cleanup can happen
+            logger.error('Waiting for canceled futures to finish')
+            wait(futures, return_when=ALL_COMPLETED)
+            ex = self.exceptions.get()
+            # TODO: What if more than one exception exists?
```
Comment (johnbradley) [inline calrissian/executor.py]: Stray TODO

File: calrissian/executor.py [hunk 3]
```diff
@@ -1,40 +1,416 @@
...
+
+        See https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
+        """
+        super(ThreadPoolJobExecutor, self).__init__()
+        self.max_workers = max_workers
+        self.jrq = JobResourceQueue()
+        self.exceptions = Queue()
+        self.total_resources = Resources(total_ram, total_cores)
+        self.available_resources = Resources(total_ram, total_cores) # start with entire pool available
+        self.resources_lock = threading.RLock()
```
Comment (johnbradley) [inline calrissian/executor.py]: Are we making use of the reentrant functionality of this lock?

File: tests/test_executor.py [hunk 4]
```diff
@@ -1,61 +1,515 @@
...
+        rsc = Resources(100, 1)
+        self.assertFalse(rsc.exceeds(self.executor.total_resources))
+        job = make_mock_job(rsc)
+        self.executor.raise_if_oversized(job)
 
+    def test_restore_raises(self):
+        large_resource = Resources(2000, 4)
+        self.assertTrue(large_resource > self.executor.total_resources)
+        # self.executor.restore(large_resource, self.logger)
+        # self.assertTrue(self.executor.available_resources < Resources.EMPTY)
```
Comment (johnbradley) [inline tests/test_executor.py]: Commented out test code.

File: Dockerfile (hunk 1 above)
Comment (dleehr) [reply Dockerfile]: Good catch, probably should do that

File: calrissian/executor.py (hunk 3 above)
Comment (dleehr) [reply calrissian/executor.py]: No - originally I did write it to use the re-entrant functionality but the allocate/restore got reworked a few times and no longer needs re-entrant locks. Good catch.

**Pull request merged**